File: runtime/go/units/wasi/wasi_filetype.go
// Map an os.FileInfo's mode to a WASI filetype tag, using Go's portable
// FileMode bits (so no platform-specific S_IFMT constants are needed).
func (w *WASI) wasi_filetype(fi os.FileInfo) byte {
    m := fi.Mode()
    switch {
    case m&os.ModeDir != 0:
        return 3 // directory
    case m&os.ModeCharDevice != 0:
        return 2 // character device
    case m&os.ModeDevice != 0:
        return 1 // block device
    case m&os.ModeSymlink != 0:
        return 7 // symbolic link
    case m&os.ModeSocket != 0:
        return 6 // socket (stream)
    case m.IsRegular():
        return 4 // regular file
    default:
        return 0 // unknown
    }
}

File: runtime/java/units/wasi/wasi_filetype.java
// Map a file's basic attributes to a WASI filetype tag. Java's portable
// BasicFileAttributes only distinguishes directory / symlink / regular / other,
// so — unlike the Go backend, which reads FileMode bits — block/character
// devices and sockets both collapse to "unknown" (0) here. That is adequate
// for the files and directories our guests touch; tty detection for the
// standard streams is handled separately in fd_fdstat_get (ADR-30).
byte wasi_filetype(java.nio.file.attribute.BasicFileAttributes a) {
    if (a.isDirectory()) {
        return 3; // directory
    }
    if (a.isSymbolicLink()) {
        return 7; // symbolic link
    }
    if (a.isRegularFile()) {
        return 4; // regular file
    }
    return 0; // unknown
}
